In [1]:
# import the necessary libraries
import pandas as pd
import numpy as np

In [2]:
# csv file with 'latin-1' encoding 
df = pd.read_csv('FoodBalanceSheets_E_Africa_NOFLAG.csv', encoding = 'latin-1')
df.head()

Area Code     Area  Item Code         Item  Element Code  \
0          4  Algeria       2501   Population           511   
1          4  Algeria       2501   Population          5301   
2          4  Algeria       2901  Grand Total           664   
3          4  Algeria       2901  Grand Total           674   
4          4  Algeria       2901  Grand Total           684   

                                  Element             Unit     Y2014  \
0           Total Population - Both sexes     1000 persons  38924.00   
1                Domestic supply quantity      1000 tonnes      0.00   
2           Food supply (kcal/capita/day)  kcal/capita/day   3377.00   
3  Protein supply quantity (g/capita/day)     g/capita/day     94.90   
4      Fat supply quantity (g/capita/day)     g/capita/day     80.06   

      Y2015     Y2016     Y2017     Y2018  
0  39728.00  40551.00  41389.00  42228.00  
1      0.00      0.00      0.00      0.00  
2   3379.00   3372.00   3341.00   3322.00  
3     94.35     94.72     92.82     91.83  
4     79.36     77.40     80.19     77.28

In [3]:
# Number 1
# arranged in order to get the highest values.
number_1 = df.groupby('Area')['Y2017'].count().sort_values(ascending = False)
print(number_1)

Area
Kenya                          1548
Egypt                          1458
Nigeria                        1457
Côte d'Ivoire                  1454
Senegal                        1452
Cameroon                       1441
Uganda                         1435
Ethiopia                       1428
Zambia                         1425
Madagascar                     1425
Malawi                         1411
United Republic of Tanzania    1409
Angola                         1407
Mozambique                     1406
Rwanda                         1397
Congo                          1395
South Africa                   1395
Guinea                         1383
Burkina Faso                   1383
Morocco                        1369
Botswana                       1368
Niger                          1365
Zimbabwe                       1365
Mauritius                      1340
Ghana                          1333
Tunisia                        1314
Chad                           1281
Sudan                  

Kenya has the highest sum with 1548.

In [4]:
# Number 5
number_5 = df[df['Area'] == 'Madagascar']
number_5.groupby('Element')['Y2015'].sum()

Element
Domestic supply quantity                  31214.98
Export Quantity                             494.75
Fat supply quantity (g/capita/day)           91.85
Feed                                       2070.22
Food                                      21120.65
Food supply (kcal/capita/day)              7685.00
Food supply quantity (kg/capita/yr)         871.59
Import Quantity                            1721.80
Losses                                     2176.00
Other uses (non-food)                      3957.12
Processing                                 1803.00
Production                                29482.89
Protein supply quantity (g/capita/day)      173.05
Residuals                                  -616.00
Seed                                        699.00
Stock Variation                            -505.00
Total Population - Both sexes             24234.00
Name: Y2015, dtype: float64

The total protein supply in Madagascar in 2015 is 173.05

In [5]:
# Number 7 --- to get the Highest Stock Variation 
number_7 = df.groupby('Element')[['Y2015', 'Y2016', 'Y2017', 'Y2018']].count()
number_7.loc['Stock Variation', :]

Y2015    4229
Y2016    4231
Y2017    4231
Y2018    4217
Name: Stock Variation, dtype: int64

The year with the highest sum of Stock Variation are 2016 and 2017, with 4231 and 4231 respectively. 

In [6]:
# Arranged from lowest to highest
number_8 = df.groupby('Area')['Y2017'].sum().sort_values()
# First seven values from the lowest 
number_8[:7]

Area
Sudan (former)               0.00
Ethiopia PDR                 0.00
Comoros                     59.84
Seychelles                 442.34
Sao Tome and Principe    12662.63
Cabo Verde               14650.74
Guinea-Bissau            19102.77
Name: Y2017, dtype: float64

Guinea Bissau is the area with the seventh lowest value of 19102.77

In [7]:
# Number_10
# year columns from 2014 - 2018
year_columns = ['Y2014', 'Y2015', 'Y2016', 'Y2017', 'Y2018']

# to calculate correlation between 'Element Code' and each year from 2014 to 2018
correlation = df[year_columns].apply(lambda year: df['Element Code'].corr(year))


# Find the year with the least correlation
# Find the year with the least correlation
least_correlated_year = correlation.idxmin()
min_correlation = correlation.min()

print(f"Year with the least correlation with 'Element Code is': {least_correlated_year}")
print(f"Correlation: {min_correlation}")    

Year with the least correlation with 'Element Code is': Y2016
Correlation: 0.02344399758496138


In [8]:
number_11 = df.describe().T
number_11

count         mean          std     min     25%      50%  \
Area Code     60943.0   134.265576    72.605709     4.0    74.0   136.00   
Item Code     60943.0  2687.176706   146.055739  2501.0  2562.0  2630.00   
Element Code  60943.0  3814.856456  2212.007033   511.0   684.0  5142.00   
Y2014         59354.0   134.196282  1567.663696 -1796.0     0.0     0.09   
Y2015         59395.0   135.235966  1603.403984 -3161.0     0.0     0.08   
Y2016         59408.0   136.555222  1640.007194 -3225.0     0.0     0.08   
Y2017         59437.0   140.917765  1671.862359 -1582.0     0.0     0.10   
Y2018         59507.0   143.758381  1710.782658 -3396.0     0.0     0.07   

                  75%       max  
Area Code      195.00     276.0  
Item Code     2775.00    2961.0  
Element Code  5511.00    5911.0  
Y2014            8.34  176405.0  
Y2015            8.46  181137.0  
Y2016            8.43  185960.0  
Y2017            9.00  190873.0  
Y2018            9.00  195875.0

The mean for Y2017 is '140.92' while the standard deviation is '1671.86'

In [9]:
# to check the number of unique values in 'Area'
number_12 = df['Area'].unique()
print('There are {} unique countries in the dataset.'.format(number_12.shape[0]))

There are 49 unique countries in the dataset.


In [10]:
number_13 = df.groupby('Element')['Y2017'].sum()
print(number_13)

Element
Domestic supply quantity                  2088198.10
Export Quantity                            182338.80
Fat supply quantity (g/capita/day)          10253.84
Feed                                       223705.68
Food                                      1258888.28
Food supply (kcal/capita/day)              454681.00
Food supply quantity (kg/capita/yr)         48690.04
Import Quantity                            294559.09
Losses                                     160614.00
Other uses (non-food)                       91645.97
Processing                                 292836.00
Production                                2030056.89
Protein supply quantity (g/capita/day)      11842.45
Residuals                                   35500.00
Seed                                        24870.14
Stock Variation                             54316.91
Total Population - Both sexes             1112641.00
Tourist consumption                            91.00
Name: Y2017, dtype: float64


The sum of processing in 2017 is 292836.00

In [11]:
# Number 15
# to create a dataframe for count of missing values and percentage missing values

missing_values = df.isnull().sum()
percentage_missing = missing_values / len(df)*100
# convert to 3 decimal places   - (3dp)
percentage_missing = round(percentage_missing, 3)

# Creating a DataFrame with two columns
missing_data = pd.DataFrame({'Missing Count': missing_values, 'Percentage Missing(%)': percentage_missing})

missing_data

Missing Count  Percentage Missing(%)
Area Code                 0                  0.000
Area                      0                  0.000
Item Code                 0                  0.000
Item                      0                  0.000
Element Code              0                  0.000
Element                   0                  0.000
Unit                      0                  0.000
Y2014                  1589                  2.607
Y2015                  1548                  2.540
Y2016                  1535                  2.519
Y2017                  1506                  2.471
Y2018                  1436                  2.356

The total missing values and percentage missing values in 2014 is 1598 and 2.607% respectively.

In [12]:
# Number 16
# the total sum of Wine produced in 2015 and 2018 respectively
grouped = df.groupby('Item')[['Y2015', 'Y2018']].sum()
wine = grouped.loc['Wine']
print(wine)

Y2015    4251.81
Y2018    4039.32
Name: Wine, dtype: float64


The total sum of wine produced in 2015 and 2018 is 4251.81 and 4039.32 respectively.

In [13]:
# Slicing Questions
number_19 = [['him', 'sell'], [90, 28, 43]]
number_19[0][1][1]

'e'

In [14]:
# to select 89, 63, 92, 48 from the the following array

array = np.array([[94, 89, 63],
                  [93, 92, 48],
                  [92, 94, 56]])

In [15]:
array[:2, 1:]

array([[89, 63],
       [92, 48]])